### ISSUE TITLE
ED Smart Contract Orders

### ISSUE THREAD

**Issue by djskinner** [etherdelta/etherdelta.github.io#299]
From the API docs:

> EtherDelta's orderbook is an off-chain mechanism (a list of cryptographically signed orders) that you can only access through this API. The API allows you to see existing orders or place a new one.

I notice that the smart contract has an order method and Order event. Is this unused/not recommended?

I'm interested for two reasons 1) testing my code on ropsten and 2) listening to the Order events directly rather than using the web socket.

Forgive me if my assumptions are wildly off. I'm new to working with Ethereum.

**Comment by zackcoburn** [comment]
You are correct that the order method in the smart contract is currently
not used. It was put in there in case we ever have a need to switch from
off-chain to on-chain orders.

That being said, you can certainly call the order function and then listen
to the event log for order events for testing purposes. The functionality
does work, it's just that orders placed in this way won't show up in the
API/UI.

**Comment by djskinner** [comment]
Cheers, I'll give the on-chain process a go for testing purposes.

**Comment by kriffo** [comment]
Zack -- please let me make sure I'm clear on this.

You've constructed a scenario where, while advertising your product as a decentralized platform, you actually handle absolutely nothing other than deposits, trades, and withdrawals on-chain.  It's entirely feasible that orders would go on-chain long before EtherDelta decides to post them to the off-chain orderbook.

Let's look at the process of posting an order.  User sends request for an order and the order, returned from the API, looks like this:

`{"id":"c5b2c0dc98b6b751f06eb8c3b834a01cd7fee77eb4def22585f0cdcda7b42267_buy","amount":"82937300000000000000","price":"0.0021112100000000096","**tokenGet**":"0xdd974d5c2e2928dea5f71b9825b8b646686bd200","**amountGet**":"82937300000000000000","**tokenGive**":"0x0000000000000000000000000000000000000000","**amountGive**":"175098057133000796","expires":"10000000000004796000","nonce":"974671385","v":28,"r":"0x33f9733fcbe5eeb07abe3c69cd40db97cbdfd9985c7fff43a9a88f1e10cb088d","s":"0x644d82e96872bd0ca78a0236bd7b598a23372b52d1af3f04dc0c5a898216cbcc","user":"**0xb41b5d74ebae52aad1a32eb11239a648ad7f83bf**","updated":"2017-12-25T20:30:23.532Z","availableVolume":"82937300000000000000","ethAvailableVolume":"82.9373","availableVolumeBase":"175098057133000800","ethAvailableVolumeBase":"0.1750980571330008","amountFilled":null},`

One might assume that this information goes on-chain, but actually none of the bold fields go on-chain.  All that goes on chain is the order hash and it's signature (so sha3(ED, tokenGet, amountGet, tokenGive, amountGive, user...)), signed using the public key of the seller (v, r, s).  Trades can't be made by only knowing the hash and its signature -- one must also know the amounts of tokens requested and offered.  So, in fact, it's in EtherDelta's best interest to make their UI and API as laggy as possible so that they have advance knowledge.  The only time literally ANYONE other than EtherDelta can fill an order is when EtherDelta decides to release the bold fields to the public -- until then, only EtherDelta can make the trade.

To summarize, a simple attack scenario would be for EtherDelta to receive the full order, post (hash(newOrder), v, r, s) to the chain, fill the order using the bold fields, and only once the order has been filled release the bold fields to the public.
